File: mirea/cources/2026-2027/aii-august/04/contest/D.cpp
#include <bits/stdc++.h>
#include <cassert>
using namespace std;
const int mod = (int)1e9+7;
// Быстрое возведение в степень:
int binpow(int a, int n) {
    int r = 1;
    while (n > 0) {
        if (n % 2 == 1)
            r = int(r * 1LL * a % mod);
        a = int(a * 1LL * a % mod);
        n >>= 1;
    }
    return r;
}
// Факториалы и обратные к ним:
const int NMAX = 1 << 21;
int fact[NMAX], ifact[NMAX];
void precalc() {
    fact[0] = 1;
    for (int i = 1; i < NMAX; i++)
        fact[i] = int(fact[i-1] * 1LL * i % mod);
    ifact[NMAX-1] = binpow(fact[NMAX-1], mod-2);
    for (int i = NMAX-2; i >= 0; i--)
        ifact[i] = int((i+1LL) * ifact[i+1] % mod);
}
// Биномиальный коэффициент:
int C(int n, int k) {
    if (n < 0 || k < 0 || k > n)
        return 0;
    return int(fact[n] * 1LL * ifact[k] % mod * ifact[n-k] % mod);
}
// Количество путей из (r1, c1) в (r2, c2):
int cntWays(int r1, int c1, int r2, int c2) {
    assert(r1 <= r2);
    if (c1 > c2) return 0;
    int dr = r2 - r1, dc = c2 - c1;
    return C(dr+dc, dr);
}
// Решение:
void solve() {
    int n, m, r1, c1, r2, c2;
    cin >> n >> m >> r1 >> c1 >> r2 >> c2;
    // отсортируем клетки так, чтобы (r1, c1) шла раньше (r2, c2)
    if (r1 > r2 || (r1 == r2 && c1 > c2)) {
        swap(r1, r2);
        swap(c1, c2);
    }
    // считаем пути по формуле включений-исключений:
    int64_t answ = cntWays(1, 1, n, m);
    answ -= cntWays(1, 1, r1, c1) * 1LL * cntWays(r1, c1, n, m);
    answ -= cntWays(1, 1, r2, c2) * 1LL * cntWays(r2, c2, n, m);
    answ += cntWays(1, 1, r1, c1) * 1LL * cntWays(r1, c1, r2, c2) % mod * cntWays(r2, c2, n, m);
    // выводим ответ:
    cout << (answ % mod + mod) % mod << '\n';
}
main() {
    precalc(); // вызов предподсчёта факториалов и обратных к ним
    int tt; cin >> tt;
    while (tt --> 0) solve();
}

File: mirea/cources/2026-2027/aii-august/04/contest/D.py
mod = int(1e9) + 7
# Факториалы и обратные к ним:
NMAX = 1 << 21
fact, ifact = [0] * NMAX, [0] * NMAX
def precalc():
    fact[0] = 1
    for i in range(1, NMAX):
        fact[i] = fact[i - 1] * i % mod
    ifact[NMAX - 1] = pow(fact[NMAX - 1], mod - 2, mod)
    for i in range(NMAX - 2, -1, -1):
        ifact[i] = (i + 1) * ifact[i + 1] % mod
# Биномиальный коэффициент:
def C(n, k):
    if n < 0 or k < 0 or k > n:
        return 0
    return fact[n] * ifact[k] % mod * ifact[n - k] % mod
# Количество путей из (r1, c1) в (r2, c2):
def cntWays(r1, c1, r2, c2):
    assert r1 <= r2
    if c1 > c2:
        return 0
    dr, dc = r2 - r1, c2 - c1
    return C(dr + dc, dr)
# Решение:
def solve():
    n, m, r1, c1, r2, c2 = map(int, input().split())
    # отсортируем клетки так, чтобы (r1, c1) шла раньше (r2, c2)
    if r1 > r2 or (r1 == r2 and c1 > c2):
        r1, r2 = r2, r1
        c1, c2 = c2, c1
    # считаем пути по формуле включений-исключений:
    answ = cntWays(1, 1, n, m)
    answ -= cntWays(1, 1, r1, c1) * cntWays(r1, c1, n, m)
    answ -= cntWays(1, 1, r2, c2) * cntWays(r2, c2, n, m)
    answ += cntWays(1, 1, r1, c1) * cntWays(r1, c1, r2, c2) % mod * cntWays(r2, c2, n, m)
    # выводим ответ:
    print((answ % mod + mod) % mod)
precalc() # вызов предподсчёта факториалов и обратных к ним
tt = int(input())
while tt > 0:
    solve()
    tt -= 1